readme.MD
```
  _   _  ____  _____  __  __ 
 | \ | |/ __ \|  __ \|  \/  |
 |  \| | |  | | |__) | \  / |
 | . ` | |  | |  _  /| |\/| |
 | |\  | |__| | | \ \| |  | |
 |_| \_|\____/|_|  \_\_|  |_|
```

NORM is an ORM that tries not to be much like an ORM.

Features
========

* Forces you to make models that uses our fields.
  * Wraps fields on your models so you know if you changed them.
  * Save only the fields that changed.
  * Only Validate fields that changed.
* Validation interface that probably does all the things.
* Barely makes queries for you, actually that may be an overstatement.
* Built on dbr, so you can make your own queries that are better than anything we could make for you.

Why?
====

Because we wanted something like an ORM, but got out of our way and let us drive. Making queries is likely still best
done by humans (for now...). Most ORM's try to do too much, and eventually block you from doing the complex business 
logic you end up needing to do 6 months from now.

Examples
========

Model
-----
```golang
type Post struct {
  Id field.Int64
  Title field.String
  Content field.String
  AuthorId field.Int64
  Created field.Time
}
```

Select a Model
--------------

```golang
var post Post = Post{}
post.Id.Scan(4384)

err := norm.NewSelect(dbrSession, post, nil).Where("id = ?", post.Id.Int64).LoadStruct(&post) // select all defined fields

err = norm.NewSelect(dbrSession, post, field.FieldNames{"id", "title"}).Where("id = ?", post.Id.Int64).LoadStruct(&post) // only those fields
```

Select Models
-------------

```golang
var posts []Post = make([]Post, 0)

err := norm.NewSelect(dbrSession, post, nil).Where("created > '2015-01-01").LoadStructs(&posts)
```

Insert Model
------------

```golang
var post Post = Post{}
post.Title.Scan("First Post")
post.Context.Scan("...")
post.AuthorId.Scan(1)
post.Created.Scan(time.Time.Now())

result, err := norm.NewInsert(dbrSession, post, nil).Exec()

fmt.Printf("InsertId: %s", result.Value())
```

Update Model
------------

```golang

// continued from Insert
post.Content.Scan("modified content")

result, err := norm.NewUpdate(dbrSession, post, nil).Exec()

```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### New
### Changed

## [0.2.4] - 2015-11-10
### Fixed
- Fixed better handling of Time.IsZero()

## [0.2.3] - 2015-11-10
### Fixed
- Better handling of Time.IsZero()

## [0.2.2] - 2015-10-27
### Fixed
- ModelValidate would return a non-nil value when it intended to return a nil value for no errors

## [0.2.1] - 2015-10-26
### Changed
- Wrapped dbr.Tx in an interface and updated Session implementation to return our wrap

## [0.2.0] - 2015-10-25
### New
- IsSet() bool added to all fields and is part of the field.Field interface
- field.Names can now return the intersection of two sets, Intersect()
### Changed
- Removed `Valid` field from non-nullable fields
- Fixed a few Marshal/Unmarshal edge cases


## [0.1.1] - 2015-10-19
### New
- Support for `ModelFields` to recurse into Anonymous/Embedded models to return all field.Names

## [0.1.0] - 2015-09-28
### New
- Connections are created with the database name, full database.tablename selectors for queries
  can be generated with `ModelTableName`.
### Changed
- Wrapped `dbr.Connection` and `dbr.Session` in our own local structs to add some context.
- Validation code refactored to two interfaces.


## [0.0.5] - 2015-09-22
- Float64 and NullFloat64 support

## [0.0.4] - 2015-09-14
### New
- Better handling of Null fields

## [0.0.3] - 2015-09-04
### New
- `ModelDirtyFields` added to get field.Names that are dirty
- `ModelLoadMap` can assign map[string]interface{} to a Model

### Changed
- Update and Insert use new PrimaryKeyer data to trim/customize queries
- JSON Marshal/Unmarshal improvements


field/time.go
package field

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"fmt"
	"gopkg.in/guregu/null.v2"
	"time"
)

const (
	timeFormat = "2006-01-02 15:04:05.000000"
)

// Time field that does not accept nil
type Time struct {
	Time   time.Time
	shadow time.Time
	ShadowInit
}

// Scan a value into the Time, error on nil or unparsable
func (t *Time) Scan(value interface{}) error {
	var err error
	if value == nil {
		t.Time = time.Time{}
	}
	switch v := value.(type) {
	case time.Time:
		if v.IsZero() {
			return ErrorCouldNotScan("Time", value)
		}
		t.Time = v
		break
	case []byte:
		t.Time, err = parseDateTime(string(v), time.UTC)
		break
	case string:
		t.Time, err = parseDateTime(v, time.UTC)
		break
	default:
		return ErrorCouldNotScan("Time", value)
	}
	if err != nil {
		t.Time = time.Time{}
		return err
	}
	// load shadow on first scan only
	t.DoInit(func() {
		t.shadow = t.Time
	})

	return nil
}

// Value return the value of this field
func (t Time) Value() (driver.Value, error) {
	return t.Time, nil
}

// ShadowValue return the initial value of this field
func (t Time) ShadowValue() (driver.Value, error) {
	if t.InitDone() {
		return t.shadow, nil
	}

	return nil, ErrorUnintializedShadow
}

// IsDirty if the shadow value does not match the field value
func (t *Time) IsDirty() bool {
	return t.Time != t.shadow
}

//IsSet indicates if Scan has been called successfully
func (t Time) IsSet() bool {
	return t.InitDone()
}

// MarshalJSON Marshal just the value of Time
func (t Time) MarshalJSON() ([]byte, error) {
	return json.Marshal(t.Time)
}

// UnmarshalJSON implements encoding/json Unmarshaler
func (t *Time) UnmarshalJSON(data []byte) error {
	tmp := null.Time{}
	err := json.Unmarshal(data, &tmp)
	if err != nil {
		return err
	}
	if tmp.Valid == false {
		return errors.New("Attempted to unmarshal null value")
	}
	return t.Scan(tmp.Time)
}

type nullTime null.Time

// NullTime time that can be nil
type NullTime struct {
	nullTime
	validNull       bool
	shadow          null.Time
	shadowValidNull bool
	ShadowInit
}

// Scan a value into the Time, error on unparsable
func (nt *NullTime) Scan(value interface{}) error {
	var err error
	switch v := value.(type) {
	case time.Time:
		if v.IsZero() {
			nt.Valid = false
			nt.validNull = true

		} else {
			nt.Time, nt.Valid = v, true
			nt.validNull = false
		}
		break
	case []byte:
		nt.Time, err = parseDateTime(string(v), time.UTC)
		if nt.Time.IsZero() == true {
			nt.Valid = false
			nt.validNull = false
			return ErrorCouldNotScan("NullTime", value)
		}
		nt.Valid = (err == nil)
		if err == nil {
			nt.validNull = false
		}
		break
	case string:
		nt.Time, err = parseDateTime(v, time.UTC)
		if nt.Time.IsZero() == true {
			nt.Valid = false
			nt.validNull = false
			return ErrorCouldNotScan("NullTime", value)
		}
		nt.Valid = (err == nil)
		if err == nil {
			nt.validNull = false
		}
		break
	default:
		if value == nil {
			nt.Valid = false
			nt.validNull = true
		} else {
			err = ErrorCouldNotScan("NullTime", value)
		}
	}

	// load shadow on first scan only
	nt.DoInit(func() {
		_ = nt.shadow.Scan(nt.Time)
		if value == nil {
			nt.shadowValidNull = true
		}
	})
	return err
}

// Value return the value of this field
func (nt NullTime) Value() (driver.Value, error) {
	if nt.validNull {
		return nil, nil
	}
	return nt.Time, nil
}

// IsDirty if the shadow value does not match the field value
func (nt *NullTime) IsDirty() bool {
	if nt.validNull && nt.shadowValidNull {
		return false
	} else if nt.validNull == false && nt.shadowValidNull == false {
		return !nt.Time.Equal(nt.shadow.Time)
	}
	return true
}

//IsSet indicates if Scan has been called successfully
func (nt NullTime) IsSet() bool {
	return nt.InitDone()
}

// ShadowValue return the initial value of this field
func (nt NullTime) ShadowValue() (driver.Value, error) {
	if nt.InitDone() {
		if nt.shadowValidNull {
			return nil, nil
		}
		return nt.shadow.Value()
	}
	return nil, ErrorUnintializedShadow
}

// MarshalJSON Marshal just the value of String
func (nt NullTime) MarshalJSON() ([]byte, error) {
	if nt.Valid {
		return json.Marshal(nt.Time)
	}
	return json.Marshal(nil)

}

// UnmarshalJSON implements encoding/json Unmarshaler
func (nt *NullTime) UnmarshalJSON(data []byte) error {
	t := &null.Time{}
	err := t.UnmarshalJSON(data)
	if err != nil {
		return err
	}
	if t.Valid == true {
		return nt.Scan(t.Time)
	}
	return nt.Scan(nil)
}

// taken from https://github.com/go-sql-driver/mysql/blob/master/utils.go
func parseDateTime(str string, loc *time.Location) (t time.Time, err error) {
	base := "0000-00-00 00:00:00.0000000"
	switch len(str) {
	case 10, 19, 21, 22, 23, 24, 25, 26: // up to "YYYY-MM-DD HH:MM:SS.MMMMMM"
		if str == base[:len(str)] {
			return
		}
		t, err = time.Parse(timeFormat[:len(str)], str)
	default:
		err = fmt.Errorf("Invalid Time-String: %s", str)
		return
	}

	// Adjust location
	if err == nil && loc != time.UTC {
		y, mo, d := t.Date()
		h, mi, s := t.Clock()
		t, err = time.Date(y, mo, d, h, mi, s, t.Nanosecond(), loc), nil
	}

	return
}
